lib: cbprintf: fix printing high-bit characters

rbrscoleman opened the pull request:
The `cbprintf_cb` callback function taken by `cbvprintf()` may return a negative number to indicate an I/O error; in the success case, it is supposed to return the (positive) input. The `str_out()` implementation can never error (it just copies the input to a buffer), but it always returns its literal input. When called with a signed character with its MSB set (e.g., UTF-8-encoded text), the integer value is sign-extended, and the function returns a negative number. This is interpreted as an error by `outs()` (in `cbprintf_complete.c`), and so printing concludes after a single character.

For example, prior to this change, the assertion in this snippet fails:

```c
char buf[16];
int len = snprintfcb(buf, sizeof(buf), "\xED\xA0\xBD\xED\xB1\x8D");
assert(len == 6);
```

This can be seen when compiling with `CONFIG_CBPRINTF_LIBC_SUBSTS=y` and `CONFIG_CBPRINTF_NANO=n`, as covered by these unit tests:

```sh
west build --pristine --board unit_testing/unit_testing tests/unit/cbprintf/ --test-item utilities.prf.m32v101
# or
west build --pristine --board unit_testing/unit_testing tests/unit/cbprintf/ --test-item utilities.prf.m64v101
```

Then try:

```sh
./build/testbinary
```

...without and with the change to `lib/os/cbprintf.c`.

rbrscoleman commented on lib/os/cbprintf.c:
```diff
@@ -45,7 +45,7 @@ static int str_out(int c,
 	}
 
-	return c;
+	return 0;
```
Ah, thank you, I missed that! Fixed now (although I used `unsigned char` rather than `uint8_t` – can change if you'd prefer).

rbrscoleman commented:
Not stale, just awaiting review again.

@nordic-krch, please do let me know if there's anything I can do to make review easier.

nordic-krch commented:
Unit tests are run on `unit_testing` virtual board. You can run it like: ` west build -p -b unit_testing/unit_testing tests/unit/cbprintf/ -T utilities.prf.m32v00` and then `./build/testbinary` to run the executable.

rbrscoleman committed:
lib: cbprintf: fix printing high-bit characters.

The `cbprintf_cb` callback function taken by `cbvprintf()` may return a
negative number to indicate an I/O error. The `str_out()` function can
never error (it just copies the input to a buffer), but it always returns
its input. For regular ASCII input, this is merely nonsense; however, when
writing out a character with its MSB set (e.g., UTF-8-encoded text), the
character is sign-extended, and the function returns a negative number.
This is interpreted as an error by `outs()` (in `cbprintf_complete.c`), and
so printing concludes after a single character.

For example, prior to this change, the assertion in this snippet fails:

    char buf[16];
    int len = snprintfcb(buf, sizeof(buf), "\xED\xA0\xBD\xED\xB1\x8D");
    assert(len == 6);

Signed-off-by: Samuel Coleman <samuel.coleman@rbr-global.com>

rbrscoleman commented:
Thanks, it was the `-T, --test-item` flag I was missing.

I've added coverage for this issue to the `prf.test_libc_substs` unit test, and I've manually checked that the test fails without the patch, and passes with.

nordic-krch commented:
you can split it into 2 commits.
